types/src/form.rs

```diff
@@ -69,21 +72,98 @@ impl AnnotatedForm {
         Ok(None)
     }
 
-    async fn romanized_source(&self) -> Option<String> {
-        self.simple_phonetics
-            .as_ref()
-            .map(|phonetic| crate::lexical::simple_phonetics_to_worcester(phonetic))
+    async fn romanized_source(&self, system: CherokeeOrthography) -> Option<Cow<'_, str>> {
+        self.simple_phonetics.as_ref().map(|phonetic| {
+            if system == CherokeeOrthography::Learner {
+                crate::lexical::simple_phonetics_to_worcester(phonetic).into()
+            } else {
+                phonetic.into()
+            }
+        })
     }
 
     async fn segments(
         &self,
         context: &async_graphql::Context<'_>,
+        system: CherokeeOrthography,
     ) -> FieldResult<Vec<MorphemeSegment>> {
-        Ok(context
-            .data::<DataLoader<Database>>()?
+        let db = context.data::<DataLoader<Database>>()?;
+        // 1. To convert to a concrete analysis, start with a list of abstract tags.
+        let abstract_segments = db
             .load_one(PartsOfWord(*self.id.as_ref().unwrap()))
             .await?
-            .unwrap_or_default())
+            .unwrap_or_default();
+
+        // 2. Request all concrete tags that start with each abstract tag.
+        let concrete_tag_matches = db
+            .load_many(
+                abstract_segments
+                    .iter()
+                    .map(|seg| TagId(seg.gloss.clone(), system)),
+            )
+            .await?;
+
+        // 3. Pick the longest match for each abstract segment.
+        let mut concrete_segments = Vec::new();
+        let mut curr_index = 0;
+        for (idx, abstract_segment) in abstract_segments.iter().enumerate() {
+            // If this segment has already been filled by a previous match, skip it.
+            if idx < curr_index {
+                continue;
+            }
+
+            let concrete_tags =
+                concrete_tag_matches.get(&TagId(abstract_segment.gloss.clone(), system));
+            if let Some(concrete_tags) = concrete_tags {
```

Comment:
That's not valid syntax in Rust afaik, are there docs detailing it?

Reply:
Maybe you were reaching for `if concrete_tags.is_some()`? I don't wanna do that because I need the inner value inside the `if` block so I bind it. What I wrote is semantically equivalent to this at runtime:

```rs
if concrete_tags.is_some() {
  let concrete_tags = concrete_tags.unwrap();
  // ...
}
```

Reply:
Got it. I think I was just looking at it like js and didn't realize the syntax.